app/src/crypto.c

```diff
@@ -22,74 +22,119 @@
-                    NULL,
-                    0);
-
-            switch (addressKind) {
-                case key_ed25519: {
-                    cx_ecfp_init_private_key(CX_CURVE_Ed25519, privateKeyData, 32, &cx_privateKey);
-                    cx_ecfp_init_public_key(CX_CURVE_Ed25519, NULL, 0, &cx_publicKey);
-                    cx_ecfp_generate_pair(CX_CURVE_Ed25519, &cx_publicKey, &cx_privateKey, 1);
-                    for (unsigned int i = 0; i < PK_LEN_25519; i++) {
-                        pubKey[i] = cx_publicKey.W[64 - i];
-                    }
-
-                    if ((cx_publicKey.W[PK_LEN_25519] & 1) != 0) {
-                        pubKey[31] |= 0x80;
-                    }
+    // Generate keys
+    switch (addressKind) {
+        case key_ed25519: {
+            ret = gen_ed25519_public_key(&cx_publicKey, path);
+            if (ret != CX_OK) {
+                return zxerr_unknown;
+            }
+
+            for (unsigned int i = 0; i < PK_LEN_25519; i++) {
+                pubKey[i] = cx_publicKey.W[64 - i];
+            }
+
+            if ((cx_publicKey.W[PK_LEN_25519] & 1) != 0) {
+                pubKey[31] |= 0x80;
+            }
+            break;
+        }
+#ifdef SUPPORT_SR25519
+        case key_sr25519: {
+            BEGIN_TRY
+            {
+                uint8_t privateKeyData[SK_LEN_25519];
+                TRY
+                {
+                    os_perso_derive_node_bip32_seed_key(
+                            HDW_NORMAL,
+                            CX_CURVE_Ed25519,
+                            path,
+                            HDPATH_LEN_DEFAULT,
+                            privateKeyData,
+                            NULL,
+                            NULL,
+                            0);
+                    get_sr25519_sk(privateKeyData);
+                    crypto_scalarmult_ristretto255_base_sdk(pubKey, privateKeyData);
                     break;
                 }
-#ifdef SUPPORT_SR25519
-                    case key_sr25519:
-                            get_sr25519_sk(privateKeyData);
-                            crypto_scalarmult_ristretto255_base_sdk(pubKey, privateKeyData);
-                        break;
-#endif
-                default:
-                    CLOSE_TRY;
-                    return zxerr_invalid_crypto_settings;
+                FINALLY
+                {
+                    MEMZERO(privateKeyData, sizeof(privateKeyData));
+                }
             }
+            END_TRY
         }
-        FINALLY
-        {
-            MEMZERO(&cx_privateKey, sizeof(cx_privateKey));
-            MEMZERO(privateKeyData, SK_LEN_25519);
-        }
+#endif
+        default:
+            return zxerr_invalid_crypto_settings;
     }
-    END_TRY;
 
     return zxerr_ok;
 }
 
+cx_err_t gen_ed25519_private_key(cx_ecfp_private_key_t* pPrivateKey, const unsigned int* path) {
+    cx_err_t ret;
+    uint8_t privateKeyData[SK_LEN_25519];
+
+    os_perso_derive_node_bip32_seed_key(
```

Comment:
This function can throw an exception. It must be enclosed in a `TRY` / `CATCH` block.

Reply:
Thanks, fixed. Note there is another unprotected call to this function in this file, but since it is part of the SR25519 scheme which may require a deeper refactor I haven't modified it.